lib/ovto/pure_component.rb

```diff
@@ -0,0 +1,20 @@
+module Ovto
+  class PureComponent < Component
+    def initialize(wired_actions)
+      super
+      @prev_props = nil
+      @cache = nil
+    end
+
+    def do_render(args, state)
+      return @cache if args == @prev_props
```

Comment:
How about using `#equal?` instead of `#==` ?

Reply:
`#equal?` does not work, because component receives different hash instance every time.

```ruby
def render
  o Pure, {foo: bar} # `{foo: bar}` hash is created by every time of `render` calling.
end
```


By the way, I guess we can use `equal?` for all value of `args`. e.g.) `args.all?{|k, v| v.equal? @prev_props[k]}`.

Reply:
I tried benchmarking, but the naive implementation to use `equal?` is slower than `==`.


note: The example in the previous comment has a bug because it does not check key equivalent.

 ```ruby
require 'benchmark'

# Use Struct because there are enough seped difference between #equal? and #==.
obj = Struct.new(:foo, :bar).new(1, 2)

a = { foo: 'bar', hoge: obj }
b = a.dup
c = { foo: 'bar', hoge: obj, additional: 'aaa' }

def use_eqeq(x, y)
  x == y
end

def use_equal(x, y)
  x.keys == y.keys && x.all?{|k, v| y[k] == v}
end


# Just test for use_equal method
p use_eqeq(a, b) # => true
p use_equal(a, b) # => true
p use_eqeq(a, c) # => false
p use_equal(a, c) # => false

N = 100000

Benchmark.bm(20) do |x|
  x.report('Struct#==')     {N.times{obj == obj}}
  x.report('Struct#equal?') {N.times{obj.equal?(obj)}}

  x.report('same ==')    {N.times{use_eqeq(a, b)}}
  x.report('same equal?'){N.times{use_equal(a, b)}}

  x.report('diff ==')    {N.times{use_eqeq(a, c)}}
  x.report('diff equal?'){N.times{use_equal(a, c)}}
end
```

```bash
$ opal test.rb
true
true
false
false
                           user     system      total        real
Struct#==              0.218000   0.218000   0.872000 (  0.217753)
Struct#equal?          0.016000   0.016000   0.064000 (  0.015879)
same ==                0.253000   0.253000   1.012000 (  0.253861)
same equal?            0.498000   0.498000   1.992001 (  0.498554)
diff ==                0.017000   0.017000   0.068000 (  0.017099)
diff equal?            0.085000   0.085000   0.340000 (  0.084590)
```


I guess we can improve speed by implementing it with JavaScript, like [Hash#==](https://github.com/opal/opal/blob/2775e1b6e3066b4cd6ee4c3c1f4df365339f7115/opal/corelib/hash.rb#L95-L127).
But I think it is over-engineering, so we can use `Hash#==`. It is enough fast.